testing/end_to_end_test.go

```diff
@@ -0,0 +1,205 @@
+	if !viper.IsSet("RESIM_CLIENT_ID") {
+		s.FailNow("RESIM_CLIENT_ID must be set")
+	}
+	if !viper.IsSet("RESIM_CLIENT_SECRET") {
+		s.FailNow("RESIM_CLIENT_SECRET must be set")
+	}
+}
+
+func (s *EndToEndTestSuite) buildCLI() string {
+	fmt.Println("Building CLI")
+	tmpDir, err := os.MkdirTemp("", TempDirSuffix)
+	s.NoError(err)
+	outputPath := filepath.Join(tmpDir, CliName)
+	buildCmd := exec.Command("go", "build", "-o", outputPath, "../cmd/resim")
+	err = buildCmd.Run()
+	s.NoError(err)
+	fmt.Println("Successfully built CLI")
+	return outputPath
+}
+
+func (s *EndToEndTestSuite) foldFlags(flags []Flag) []string {
+	flagsSlice := []string{}
+	for _, flag := range flags {
+		flagsSlice = append(flagsSlice, flag.Name)
+		flagsSlice = append(flagsSlice, flag.Value)
+	}
+	return flagsSlice
+}
+
+func (s *EndToEndTestSuite) buildCommand(commandBuilders []CommandBuilder) *exec.Cmd {
+	// We populate the URL and the auth URL flags initially, then for each command/flags pair
+	// in the command builder, we generate a mega slice:
+	allCommands := []string{"--url", s.Endpoint}
+	allCommands = append(allCommands, []string{"--auth-url", s.Config.AuthKeyProviderDomain}...)
+	for _, commandBuilder := range commandBuilders {
+		allCommands = append(allCommands, commandBuilder.Command)
+		for _, flag := range s.foldFlags(commandBuilder.Flags) {
+			allCommands = append(allCommands, flag)
+		}
+	}
+	fmt.Println("Command: ", allCommands)
+	return exec.Command(s.CliPath, allCommands...)
+}
+
+func (s *EndToEndTestSuite) runCommand(commandBuilders []CommandBuilder) Output {
+	var stdout, stderr bytes.Buffer
+	cmd := s.buildCommand(commandBuilders)
+	fmt.Println("Full command: ", cmd.String())
+	cmd.Stdout = &stdout
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+	fmt.Println("Out: ", stdout.String())
+	fmt.Println("Error: ", stderr.String())
+	s.NoError(err)
+	return Output{
+		StdOut: stdout.String(),
+		StdErr: stderr.String(),
+	}
+}
+
+// As a first test, we expect the help command to run successfully
+func (s *EndToEndTestSuite) TestHelp() {
+	fmt.Println("Testing help command")
+	runCommand := CommandBuilder{
+		Command: "help",
+	}
+	output := s.runCommand([]CommandBuilder{runCommand})
+	s.Contains(output.StdOut, "Usage:")
+}
+
+func (s *EndToEndTestSuite) TestProjectCreate() {
+	fmt.Println("Testing project create command")
+
+	// We build a create project command with the name and description flags
+	projectCommand := CommandBuilder{
+		Command: "project",
+	}
+	createCommand := CommandBuilder{
+		Command: "create",
+		Flags: []Flag{
+			{
+				Name:  "--name",
+				Value: "test-project",
+			},
+			{
+				Name:  "--description",
+				Value: "description",
+			},
+		},
+	}
+	output := s.runCommand([]CommandBuilder{projectCommand, createCommand})
+	fmt.Println("Output: ", output.StdOut)
+	s.Contains(output.StdOut, "Created project")
+}
+
+func TestEndToEndTestSuite(t *testing.T) {
+	viper.AutomaticEnv()
```

Comment:
Note: i am not using a prefix here because it messes with passing in the RESIM_CLIENT_ID etc